### DISCUSSION TITLE
How do you keep an ESP32 AP running 24/7?

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#14436] [ESP32] [answered]
I am trying to running a simple server on an ESP32 AP that listens for a few clients.  I think I have ways of telling if the server part is running, however the system locks-up every day or so.

Is there a way for the ESP32 itself to tell if the AP is running and then if not ... do a machine.reset()?? 

Would running asyncio with a server task and another task blinking an LED with a watchdog timer be able to tell me if the server parts locks-up?

This what I am running as the server:
```py
while True:
    try:
     #  allow for immediate re-use
        server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server.bind((config.host, config.port))
    except:
        socket_count += 1

        try:
            with open('errors.txt', 'a') as outfile:
                outfile.write('something failed at the socket stage' + '\n')
                outfile.write('socket fail count' + str(socket_count) + '\n')
        except OSError:
            pass

        machine.reset()

    while True:
        try:
            server.listen()
            client, address = server.accept()
            break
        except:
            listen_count += 1

            try:
                with open('errors.txt', 'a') as outfile:
                    outfile.write('something failed at the listen stage' + '\n')
                    outfile.write('listen fail count' + str(listen_count) + '\n')
            except OSError:
                pass

  # try starting again if listen fails
    break
```

**Comment by glenn20** [comment, marked as answer]
One approach is to have a client which is expected to send heartbeat signals to your server. In this way, it doesnt matter what state your software/hardware gets into, but it does rely on at least one active heartbeat device on your network.

You could use the heartbeat to feed a watchdog or have your watchdog task check the time since the last heartbeat and stop feeding the watchdog if the heartbeat has timed out.

**Reply by davefes** [reply]
Thanks @glenn20.  I recalled your ping examples in ESPNow but could not work-out how that was going re-boot the server.  Now I understand.

I would prefer that the clients (solar-powered remotes) were not pinging all the time, so maybe another local client that can run off mains could be a solution.  I will review your ping example.

**Comment by heiqiaokeli1947** [comment]
use microdot as webserver.
https://microdot.readthedocs.io/en/latest/index.html

**Reply by davefes** [reply]
I have considered the "jump" in complexity to try Microdot.  For several years now my biggest implementation challenge has been trying to keep ESP32s connected to WiFi for long periods of time.  Connecting only when required and watchdog timers seem to help.

I thought I would make my own internal WiFi network to collect all the data I want then connect to the internet to transfer datalogs, alarms and images.  So, my AP just listens 24/7 for clients.  I can accept the fact that the actual communications link could fail, but not that a simple listening server using the code above should only run for a day or so without intervention.  On other projects I have "listened" for say 1-10 minutes using `timeout`, run a watchdog timer and then go back to listening.  I don't know if I would miss client requests while I was servicing the watchdog. 

Peter Hinch's robust WiFi approach
https://github.com/peterhinch/micropython-iot/blob/master/README.md#7-quality-of-service
would improve the link-side of things.  glenn20s's suggestion above could improve the overall system ... just need to ensure that the "pinging" client keeps working!  

If you can tell me why Microdot would be more robust ... I am eager to learn.

**Reply by davefes** [reply]
Hi @glenn20,

Found a handy article here:
https://RandomNerdTutorials.com/raspberry-pi-pico-w-asynchronous-web-server-micropython/
so I modified it to use as a "pinged" test server.

As follows:
```py
import network
import asyncio
import socket
import time
from wifi_ap_sta import connect_ap, disconnect_ap, connect_sta, disconnect_sta
from machine import Pin
import gc
from watchdog import wdt_init, wdt_feed


gc.collect()

# SW/WDT set for 70 seconds
wdt_init(70)
wdt_feed()


led_blink = Pin(13, Pin.OUT)  # blink led to show uasyncio is non-blocking


# asynchronous function to handle client's requests
async def handle_client(reader, writer):
    print('Client connected')
#    request_line = await reader.readline()
#    print('Request:', request_line)

    data = await reader.read(32)
    message = data.decode()
#    addr = writer.get_extra_info('peername')
#    print(f"Received {message!r} from {addr!r}")

    if message == 'ping':
        wdt_feed()

    print(f"Send: {message!r}")
    writer.write(data)
    await writer.drain()

    print("Close the connection")
    writer.close()
    await writer.wait_closed()


async def blink_led():
    while True:
        led_blink.value(not led_blink.value())  # toggle led, esp32 pin has no toggle
        print('blink')
        await asyncio.sleep(0.5)  # blink interval


async def main():
    print('starting as an AP...')
    connect_ap()

    # start the server and run the event loop
    print('Setting up server')
    server = asyncio.start_server(handle_client, '192.168.4.1', 10000)
    asyncio.create_task(server)
    asyncio.create_task(blink_led())

# create an Event Loop
loop = asyncio.get_event_loop()
# create a task to run the main function
loop.create_task(main())

try:
    # run the event loop indefinitely
    loop.run_forever()
except Exception as e:
    print('Error occured: ', e)
except KeyboardInterrupt:
    print('Program Interrupted by the user')
```

This runs properly, but reading through:
https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md
it seems to be better practice to not use event loops.  I had a go, but was
unsuccessful.

Found another example that said to use:
`await server.start_serving()`
but Micropython does not seem to like it.

I also put a watchdog in the client so here's hoping this will keep
my system alive 24/7.

Thanks again for the hint.

**Reply by glenn20** [reply]
@davefes Great.

On avoiding explicit usage of the `loop`, I *believe* the following should work... (replace the last lines of your code).

```py
    server = await asyncio.start_server(handle_client, '192.168.4.1', 10000)
    asyncio.create_task(blink_led())
    await server.wait_closed()  # micropython does not have `serve_forever()`

asyncio.run(main())
```

Dont forget the `await` before `start_server()` will create the task to listen for incoming connections. Of course, you can also wrap the last line in a try...except clause as in your code.

The nice thing about embedding the heartbeat and wd in your app code is that you are testing the full chain. A limitation is that you get no diagnostic information about what failed before resetting the device. I typically run two heartbeat devices on the network and also set those devices to reset using a wd if they are unable to connect to any expected servers.

**Reply by davefes** [reply]
Bingo, it works.  I had everything except the `wait_closed`.  Appears to be an adequate replacement for `start_serving', even though they don't sound like similar methods.

Now, I will have to find the whole asyncio API!

We'll see how one heartbeat device improves things, before doubling-up.  Thanks again,

**Reply by glenn20** [reply]
> Now, I will have to find the whole asyncio API!

I generally start from the official python api docs to get the python best practice (eg. the  [`start_server docs`](https://docs.python.org/3/library/asyncio-stream.html#tcp-echo-server-using-streams)) and then check for coverage by the available micropython methods at [micropython asyncio docs](https://docs.micropython.org/en/latest/library/asyncio.html#asyncio.Server) and also in the [code](https://github.com/micropython/micropython/blob/master/extmod/asyncio/stream.py#L127) when necessary. And, of course, Peter Hinch's brilliant asyncio tutorials (though I dont believe he covers asyncio tcp).

**Reply by davefes** [reply]
I had briefly read the first two links, but not the code.  I need a basic understanding before digging-down.  Have read Peter's tutorials several times, but I generally find that actually trying to get something to work motivates me more.  Up until this server I have never had to use asyncio.

**Reply by davefes** [reply]
Next problem ... the server accepts 9 pings then the watchdog timer kicks in at 90 seconds.  The pings can be at 5 seconds or 60 seconds and the same thing happens, so I think the problem is at the server.

On my non-asyncio server I had a line:
`server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)`
that seemed to fix rapid reuse problems.

I searched `REUSEADDR asyncio` and found:
https://stackoverflow.com/questions/55879847/asyncio-how-to-reuse-a-socket
where it is suggested that:
`asyncio.open_connection(server, port)`
could be used.  However, looking at the rest of the example I am now lost again.

I suspect that reusing the address may not the problem, as I read that reuse might be automatic on some connections.

Any suggestions?

**Reply by glenn20** [reply]
Hmmm... I dont think you are exposed to rapid socket re-use issues here (if I understand your app workflow). If you are rebooting your server, rather than attempting to just restart your service repeatedly, then you won't trigger issues with trying to re-open closed sockets which are in a TIME_WAIT state.

If you know that your server is receiving the pings as expected, but your device is still rebooting after the wd timeout, that suggests something is wrong with the wd or it is not getting the feed when the pings are received. Are you using the [hardware watchdog](https://docs.micropython.org/en/latest/library/machine.WDT.html)?

**Reply by davefes** [reply]
I think the plan is to only re-boot the server when something goes wrong.
Here is what I see at the server:
```
MPY: soft reboot
you got a SOFT_RESET, so do a machine.reset()
ets Jul 29 2019 12:21:46

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4728
load:0x40078000,len:14888
load:0x40080400,len:3368
entry 0x400805cc
wait 2 seconds for a CTRL-C
starting as an AP...
access point config: ('192.168.4.1', '255.255.255.0', '192.168.4.1', '192.168.4.1')
Setting up server
Client connected
Received 'ping' from ('192.168.4.85', 56770)
Client connected
Received 'ping' from ('192.168.4.85', 56771)
Client connected
Received 'ping' from ('192.168.4.85', 56772)
Client connected
Received 'ping' from ('192.168.4.85', 56773)
Client connected
Received 'ping' from ('192.168.4.85', 56774)
Client connected
Received 'ping' from ('192.168.4.85', 56775)
Client connected
Received 'ping' from ('192.168.4.85', 56776)
Client connected
Received 'ping' from ('192.168.4.85', 56777)
Client connected
Received 'ping' from ('192.168.4.85', 56778)
ets Jul 29 2019 12:21:46

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
```
I don't see the 10th ping as sent, so I don't know if it is really received.  I will put in some more print statements.

I am using the software watchdog that you modified for me about a year ago.  I will change to the WDT in `machine`.

**Reply by davefes** [reply]
Ah, in:
```py
async def handle_client(reader, writer):
    print('Client connected')

    data = await reader.read(32)  # just for the message
    message = data.decode()
    addr = writer.get_extra_info('peername')

    print(f"Received {message!r} from {addr!r}")

    if message == 'ping':
        print('you should get a ping and then pat the dog')
        wdt_feed()
```
I don't see the 10th message. So the ping client keeps sending pings but the server seems to stop listening.

**Reply by davefes** [reply]
Behaves the same with the WDT in `machine`.

**Reply by glenn20** [reply]
So, are you saying that the device reboots (due to watchdog timing out) before it can receive the 10th ping OR that it receives the 9th ping then reboots 70 seconds later because it stopped receiving the pings? ie. is the wd firing when it should not, OR does the server stop receiving the messages, so the wd fires as it should?

**Reply by davefes** [reply]
The server stops receiving messages, so the wdt fires as it should.  With machine WDT:
```
E (167664) task_wdt: Task watchdog got triggered. The following tasks/users did not reset the watchdog in time:
E (167664) task_wdt:  - mpy_machine_wdt
E (167664) task_wdt: Tasks currently running:
E (167664) task_wdt: CPU 0: IDLE
E (167664) task_wdt: CPU 1: IDLE
E (167664) task_wdt: Print CPU 0 (current core) backtrace


Backtrace: 0x40149E9F:0x3FFBFE00 0x40083061:0x3FFBFE30 0x4009878F:0x3FFBC220 0x4014846A:0x3FFBC240 0x400D3083:0x3FFBC260 0x401ED853:0x3FFBC280

E (167664) task_wdt: Aborting.
E (167664) task_wdt: Print CPU 1 backtrace
```
I tried re-booting the ping client after the 9th ping and when it sent more pings they were not received.  Something magic about the 10th try?

**Reply by glenn20** [reply]
Ah - sorry. I thought you were saying the device was rebooting before the 10th ping was sent. Sorry for the wild goose chase.

**Reply by davefes** [reply]
Let's OK.

Here is a minimal non-working example, in case I missed something:
```py
import network
import asyncio
import socket
import time
from machine import Pin, WDT
import gc
import machine
from wifi_ap_sta import connect_ap


gc.collect()

led_blink = Pin(13, Pin.OUT)

#  system WDT set for 90 seconds
wdt = WDT(timeout = 90000)  # enable it with a 90 second timeout
wdt.feed()


async def handle_client(reader, writer):
    print('Client connected')

    data = await reader.read(32)  # just for the message
    message = data.decode()
 
    if message == 'ping':
        print('got a ping')
        wdt.feed()


async def blink_led():
    while True:
        led_blink.value(not led_blink.value())  # toggle led, esp32 pin has no toggle
        await asyncio.sleep(1)                  # blink interval


async def main():
    print('starting as an AP...')
    connect_ap()

  # start the server
    print('Setting up server')
    server = await asyncio.start_server(handle_client, '192.168.4.1', 10000)
    asyncio.create_task(blink_led())
    await server.wait_closed()

asyncio.run(main())
```

**Reply by davefes** [reply]
@glenn20  Is your comment in this discussion
https://github.com/orgs/micropython/discussions/13592
relevant?

**Reply by davefes** [reply]
Seems relevant as the following appears to keep running:
```py
import socket
from wifi_ap_sta import connect_ap


async def handle_echo(reader, writer):
    data = await reader.read(100)
    message = data.decode()
    addr = writer.get_extra_info('peername')

    print(f"Received {message!r} from {addr!r}")

    print(f"Send: {message!r}")
    writer.write(data)
    await writer.drain()

    print("Close the connection")
    writer.close()
    await writer.wait_closed()

async def main():
    print('starting as an AP...')
    connect_ap()

    server = await asyncio.start_server(
        handle_echo, '192.168.4.1', 10000)

    async with server:
        while True:
            await asyncio.sleep(300)

asyncio.run(main())
```
Thanks for posting that hint.

**Reply by glenn20** [reply]
Good detective work. It would suggest that `Serve.wait_closed()` may not be working as expected. I believe it is supposed to wait until the server task (started by `start_server()`) is closed (which should never happen in this case). Perhaps it is terminating prematurely for some reason.

If so, you could return to using `wait_closed()` and add a print statement afterward to see if the `main()` function is terminating prematurely. 

It is worth noting that microdot uses `start_server()` and `wait_closed()`, so i would expect it to work correctly.

**Reply by glenn20** [reply]
Thinking.... It might be caused by calling `wait_closed()` too soon after `serve_forever()` so that the server task is not fully initialised before trying to `await` the task. I'm intrigued and will run a few tests. But, for now, you have put together some woirking some working code :).

**Reply by davefes** [reply]
If I do:
```py
async def main():
    print('starting as an AP...')
    connect_ap()

  # start the server
    print('Setting up server')
    server = await asyncio.start_server(handle_client, '192.168.4.1', 10000)
    asyncio.create_task(blink_led())

#    async with server:
#        while True:
#            await asyncio.sleep(300)
    print('now do server.wait_closed')
    await server.wait_closed()
    print('did server.wait closed')

asyncio.run(main())
```
It runs continuously, even if I comment out the last print statement.

**Reply by glenn20** [reply]
Hmmm. In my own tests, I have not been able to replicate the issues you were seeing (using mpy v1.22.1). I was speculating that possibly the `wait_closed()` was being called before the task was fully initialised and may have been returning immediately. And eventually the server object would be garbage collected causing the server task to terminate (or some other unexpected behaviour).

If you were to delete/comment the `print('now do server.wait_closed')`, then you **should** see the issues arise again.

**Reply by davefes** [reply]
An hour ago I verified that removing the print statement caused the error.

Doing it again the server now runs continuously ?!?

**Reply by davefes** [reply]
MicroPython v1.22.2 on 2024-02-22; Generic ESP32 module with SPIRAM with ESP32

I could try a Generic without SPIRAM.

**Reply by davefes** [reply]
MicroPython v1.22.2 on 2024-02-22; Generic ESP32 module with ESP32

And it just keeps running!  The reason I wanted to use the SPIRAM variant is that when pictures and datalogs don't find their way to Gmail I save them locally.  Wanted more room.

I'll do some stress testing with:
```py
async def main():
    print('starting as an AP...')
    connect_ap()

  # start the server
    print('Setting up server')
    server = await asyncio.start_server(handle_client, '192.168.4.1', 10000)
    asyncio.create_task(blink_led())
    await server.wait_closed()

asyncio.run(main())
```

**Reply by davefes** [reply]
Maybe `connect_ap()` should only be done once before running main.
